File: backend/rating/src/main/java/com/commonsengine/rating/service/RatingService.java
package com.commonsengine.rating.service;

import com.commonsengine.rating.domain.Model.CreditProfile;
import com.commonsengine.rating.domain.Model.Rating;
import com.commonsengine.rating.domain.Model.RatingTag;
import com.commonsengine.rating.infrastructure.persistence.RatingMapper;
import com.commonsengine.rating.infrastructure.persistence.RatingRepository;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.ArrayList;
import java.util.EnumMap;
import java.util.List;
import java.util.Map;

/**
 * 信用评价服务。
 *
 * <p>核心原则：
 * <ol>
 *   <li>双向评价——劳动者也能评价消费者</li>
 *   <li>反惩罚——评价不挂钩接单资格，仅作参考</li>
 *   <li>数据可携带——劳动者可导出自己的信用记录</li>
 * </ol>
 *
 * <p>持久化：使用 JPA + PostgreSQL，重启不丢数据。
 */
@Service
public class RatingService {

    private final RatingRepository repository;

    public RatingService(RatingRepository repository) {
        this.repository = repository;
    }

    /** 提交评价 */
    @Transactional
    public Rating submit(Rating rating) {
        repository.save(RatingMapper.toEntity(rating));
        return rating;
    }

    /** 查询某人收到的所有评价 */
    @Transactional(readOnly = true)
    public List<Rating> findReceived(String memberId) {
        return repository.findByRateeId(memberId).stream()
                .map(RatingMapper::toDomain)
                .toList();
    }

    /** 查询某人发出的所有评价 */
    @Transactional(readOnly = true)
    public List<Rating> findGiven(String memberId) {
        return repository.findByRaterId(memberId).stream()
                .map(RatingMapper::toDomain)
                .toList();
    }

    /** 查询某笔交易的评价（双方向） */
    @Transactional(readOnly = true)
    public List<Rating> findByTransaction(String transactionId) {
        return repository.findByTransactionId(transactionId).stream()
                .map(RatingMapper::toDomain)
                .toList();
    }

    /**
     * 聚合信用画像。
     */
    @Transactional(readOnly = true)
    public CreditProfile getCreditProfile(String memberId) {
        List<Rating> received = findReceived(memberId);
        if (received.isEmpty()) {
            return new CreditProfile(memberId, 5.0, 0, Map.of(), List.of());
        }

        List<Integer> scores = received.stream().map(Rating::score).toList();
        double avg = scores.stream().mapToInt(Integer::intValue).average().orElse(5.0);

        Map<RatingTag, Integer> tagFreq = new EnumMap<>(RatingTag.class);
        for (Rating r : received) {
            for (RatingTag tag : r.tags()) {
                tagFreq.merge(tag, 1, Integer::sum);
            }
        }

        // 最近 20 条评分
        List<Integer> recent = new ArrayList<>(scores);
        int from = Math.max(0, recent.size() - 20);
        List<Integer> recentScores = recent.subList(from, recent.size());

        return new CreditProfile(memberId, avg, received.size(), tagFreq, recentScores);
    }

    /**
     * 导出信用记录（数据归个人——劳动者可带走）。
     */
    @Transactional(readOnly = true)
    public String exportProfile(String memberId) {
        CreditProfile profile = getCreditProfile(memberId);
        List<Rating> received = findReceived(memberId);

        StringBuilder sb = new StringBuilder();
        sb.append("=== 公地引擎 · 信用记录导出 ===\n");
        sb.append("成员 ID: ").append(profile.memberId()).append('\n');
        sb.append("平均评分: ").append(String.format("%.2f", profile.averageScore())).append('\n');
        sb.append("评价总数: ").append(profile.totalRatings()).append('\n');
        sb.append('\n');
        sb.append("标签统计:\n");
        profile.tagFrequency().forEach((tag, count) ->
                sb.append("  ").append(tag.name()).append(": ").append(count).append(" 次\n"));
        sb.append('\n');
        sb.append("评价明细:\n");
        for (Rating r : received) {
            String tags = r.tags().stream().map(Enum::name).reduce((a, b) -> a + "," + b).orElse("");
            sb.append("  [").append(r.createdAt())
                    .append("] 分数:").append(r.score())
                    .append(" 标签:").append(tags)
                    .append(' ').append(r.comment() == null ? "" : r.comment())
                    .append('\n');
        }
        sb.append('\n');
        sb.append("注：此信用记录归您个人所有，可携带至其他合作社。\n");
        return sb.toString();
    }
}

File: backend/rating/src/main/kotlin/com/commonsengine/rating/service/RatingService.kt
package com.commonsengine.rating.service

import com.commonsengine.rating.domain.CreditProfile
import com.commonsengine.rating.domain.Rating
import com.commonsengine.rating.domain.RatingId
import com.commonsengine.rating.infrastructure.persistence.RatingRepository
import com.commonsengine.rating.infrastructure.persistence.toDomain
import com.commonsengine.rating.infrastructure.persistence.toEntity
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

/**
 * 信用评价服务
 *
 * 核心原则：
 * 1. 双向评价——劳动者也能评价消费者
 * 2. 反惩罚——评价不挂钩接单资格，仅作参考
 * 3. 数据可携带——劳动者可导出自己的信用记录
 *
 * 持久化：使用 JPA + PostgreSQL，重启不丢数据。
 */
@Service
open class RatingService(
    private val repository: RatingRepository,
) {

    /** 提交评价 */
    @Transactional
    open fun submit(rating: Rating): Rating {
        repository.save(rating.toEntity())
        return rating
    }

    /** 查询某人收到的所有评价 */
    @Transactional(readOnly = true)
    open fun findReceived(memberId: String): List<Rating> =
        repository.findByRateeId(memberId).map { it.toDomain() }

    /** 查询某人发出的所有评价 */
    @Transactional(readOnly = true)
    open fun findGiven(memberId: String): List<Rating> =
        repository.findByRaterId(memberId).map { it.toDomain() }

    /** 查询某笔交易的评价（双方向） */
    @Transactional(readOnly = true)
    open fun findByTransaction(transactionId: String): List<Rating> =
        repository.findByTransactionId(transactionId).map { it.toDomain() }

    /**
     * 聚合信用画像
     */
    @Transactional(readOnly = true)
    open fun getCreditProfile(memberId: String): CreditProfile {
        val received = findReceived(memberId)
        if (received.isEmpty()) {
            return CreditProfile(
                memberId = memberId,
                averageScore = 5.0,
                totalRatings = 0,
                tagFrequency = emptyMap(),
                recentScores = emptyList(),
            )
        }

        val scores = received.map { it.score }
        val tagFreq = received.flatMap { it.tags }.groupingBy { it }.eachCount()

        return CreditProfile(
            memberId = memberId,
            averageScore = scores.average(),
            totalRatings = received.size,
            tagFrequency = tagFreq,
            recentScores = scores.takeLast(20),  // 最近 20 条
        )
    }

    /**
     * 导出信用记录（数据归个人——劳动者可带走）
     */
    @Transactional(readOnly = true)
    open fun exportProfile(memberId: String): String {
        val profile = getCreditProfile(memberId)
        val received = findReceived(memberId)

        return buildString {
            appendLine("=== 公地引擎 · 信用记录导出 ===")
            appendLine("成员 ID: ${profile.memberId}")
            appendLine("平均评分: ${"%.2f".format(profile.averageScore)}")
            appendLine("评价总数: ${profile.totalRatings}")
            appendLine()
            appendLine("标签统计:")
            profile.tagFrequency.forEach { (tag, count) ->
                appendLine("  ${tag.name}: $count 次")
            }
            appendLine()
            appendLine("评价明细:")
            received.forEach { r ->
                appendLine("  [${r.createdAt}] 分数:${r.score} 标签:${r.tags.joinToString(",") { it.name }} ${r.comment ?: ""}")
            }
            appendLine()
            appendLine("注：此信用记录归您个人所有，可携带至其他合作社。")
        }
    }
}
